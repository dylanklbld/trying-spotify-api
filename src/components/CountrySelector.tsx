import React, { ReactElement, useState } from 'react'
import styled, { keyframes } from "styled-components";

const Container = styled.div`
  width: 90%;
  margin: 0 auto;
  display: flex;
  flex-flow: row;
  justify-content: center;
  text-align: center;
  font-family: "Courier New", Courier, monospace;
  color: white;
`

const FlagButton = styled.button<{isActive?:boolean}>`
  font-size:2em;
  cursor:pointer;
  padding: 1px 8px;
  border-radius: 3px;
  color: #17323F;
  opacity:${props=>props.isActive ? 0.5 : 1}
`;

type Props = {
   countryMap: any
   setCountry: (countryCode: string | null) => void,
   selected?: string | null
}

const resetText = "ðŸ”„"

const CountrySelector = ({ countryMap, setCountry, selected }: Props):ReactElement => {
    return (<Container>
        {Object.keys(countryMap).map((countryCode)=>{
            return <FlagButton key={countryCode} 
            onClick={()=>setCountry(countryCode)} 
            isActive={selected===countryCode}>
                {countryMap[countryCode]}
            </FlagButton>
        })}
        <button onClick={()=>setCountry(null)}>{resetText}</button>
      </Container>)
}

export default CountrySelector