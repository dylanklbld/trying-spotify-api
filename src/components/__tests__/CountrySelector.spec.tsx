import { fireEvent, render, screen } from '@testing-library/react';

import CountrySelector from '../CountrySelector'
import React from 'react';
import styled from 'styled-components';

const mockHistoryPush = jest.fn();
const mockUrlHash =`#access_token=My123Token&token_type=Bearer&expires_in=3600`

const countryMap ={
  'DE': 'Germany',
  'RU': 'Russia'
}

const setCountryMock = jest.fn().mockName('setCountry')

describe('CountrySelector', () => {
  it('should render', () => {
    const {container} = render(<CountrySelector {...{countryMap, setCountry:setCountryMock}}/>)
    expect(container).toMatchSnapshot()
  })
  
  it('should render with selected country', () => {
    const {container} = render(<CountrySelector {...{countryMap, setCountry:setCountryMock, selected:"RU"}}/>)
    
    const selectedButton = screen.getByText('Russia')
    const style1 = window.getComputedStyle(selectedButton)
    expect(style1.opacity).toBe("0.5")


    const otherButton = screen.getByText('Germany')
    const style2 = window.getComputedStyle(otherButton)
    expect(style2.opacity).toBe("1")


    expect(container).toMatchSnapshot()
  })

  it('should call select country', () => {
    render(<CountrySelector {...{countryMap, setCountry:setCountryMock, selected:"DE"}}/>)
    const button = screen.getByText('Russia')
    fireEvent.click(button)

    expect(setCountryMock).toBeCalledWith("RU")
  })

  it('should call reset', () => {
    render(<CountrySelector {...{countryMap, setCountry:setCountryMock, selected:"DE"}}/>)
    const resetButton = screen.getByText("ðŸ”„")
    fireEvent.click(resetButton)

    expect(setCountryMock).toBeCalledWith(null)
  })
})