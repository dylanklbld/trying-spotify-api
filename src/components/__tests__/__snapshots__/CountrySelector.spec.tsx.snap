// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelector should render 1`] = `
<div>
  <div
    class="sc-bdfBwQ kyTowY"
  >
    <button
      class="sc-gsTCUz kgbfGE"
    >
      Germany
    </button>
    <button
      class="sc-gsTCUz kgbfGE"
    >
      Russia
    </button>
    <button>
      ðŸ”„
    </button>
  </div>
</div>
`;

exports[`CountrySelector should render with selected country 1`] = `
<div>
  <div
    class="sc-bdfBwQ kyTowY"
  >
    <button
      class="sc-gsTCUz kgbfGE"
    >
      Germany
    </button>
    <button
      class="sc-gsTCUz cbhDZi"
    >
      Russia
    </button>
    <button>
      ðŸ”„
    </button>
  </div>
</div>
`;
